draft-ietf-taps-interface.md

```diff
@@ -646,7 +648,24 @@ In case a unidirectional connection is requested, but unidirectional connections
 the system should fall back to bidirectional transport.
 
 
-### Timeout for aborting Connection Establishment {#conn-establish-timeout}
+### Notification of excessive retransmissions {#prop-establish-retrans-notify}
+
+This property specifies whether an application considers it useful to be
+informed in case sent data was retransmitted more often than a certain
+threshold. The recommended default is to have this option.
+
+
+### Notification of ICMP soft error message arrival {#prop-soft-error}
+
+This property specifies whether an application considers it useful to be
+informed when an ICMP error message arrives that does not force termination of a
+connection. When set to true, received ICMP errors will be available as
+SoftErrors. Note that even if a protocol supporting this property is selected,
+not all ICMP errors will necessarily be delivered, so applications cannot rely
+on receiving them. The recommended default is not to have this option.
```

Comment:
As ICMP is unreliable anyway, the remark is somewhat redundant.

Reply:
There could be a perception issue here, because we see this differently.

* First, transports ignore soft errors (as specified by the IETF), primarily because after years of experience this was found to be the correct thing to do. Sure this extra information *could* help in some cases, it can also be very unhelpful in many cases (one obvious case is where there is some parallelism and packets are forwarded across more than one set of nodes - e.g. after a failover,  ... when the path changes, this is made worse when there are delayed messages due to intermittent forwarding ). I'd suggest the end to argument identifies these messages are for diagnostics, not to influence the forwarding/transport.

(2) I think you suggest the logic does not exist to forward this presently through the API. That is as I expected, and adding this as a requirement to TAPS is adding another requirement to the stack developer, TCP implementor, etc - all of which seems like a hurdle to deploying TAPS rather than a selling point. Many UDP API's don't even signal hard errors.

Based on this, I argue soft errors from the network (ICMP) must not be passed to an application. Soft errors from the transport (such as retry failures, timeouts) could be passed if there is a need.

Gorry

Reply:
@gorry - I'm not sure you answered in the right place - I think you're answering my comment, not Phil's?  Or maybe it is Phil's, because of your point (2), where I don't understand why you say "I think you suggest the logic does not exist to forward this presently through the API. "

Either way: your argument of not passing ICMP messages to the application at all makes sense to me, yet it's a divergence from minset and as such not offering everything that TCP and UDP offer. In a way it's quite a serious divergence, as we're stepping away from RFC 1122 here.

An option I see is to take this discussion to the mailing list, and if there's agreement to do this, do it and capture the reasoning in one of our documents.

Reply:
P.S. I just re-read RFC 5461, and I like the phrase:

 "When the connection timer expires and an ICMP soft error message has
   been received before the timeout, TCP can use this information to
   provide the user with a more specific error message (see [Stevens],
   pp. 317-319)."

P.P.S It wouldn't be a deployment hurdle, if it were optional to implement;-).

Reply:
I like that phrase too but I'm not sure this changes anything about the API... it's a TCP-internal thing. I like the idea of going optional !

Reply:
I agree that _requiring_ an TAPS implementation to forward soft errors is a deployment hurdle, but I also think it is a useful feature. 
What I would like to see is _requiring the support of the property_ but making the functionality optional (so the property can be a NO-OP). I am fine with adding a warning against relying on ICMP generating soft errors.

Reply:
@philsbln @gorryfair I wanted to update the text to address your comment now, but then I saw that the text about "Soft Errors" already says:

"This will only happen if the underlying protocol stack supports
access to soft errors; however, even if the underlying stack supports it, there
is no guarantee that a soft error will be signaled."

The part after the semicolon is what you want, right?

So ... all is optional. Which text changes are you then asking for?

Reply:
@mwelzl I am fine with the text for soft errors. 

We should just add a general remark about optional / NO-OP selection properties in a separate PR.